cmd/contour/serve.go

```diff
@@ -172,35 +175,45 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 	// step 4. register our resource event handler with the k8s informers.
 	coreInformers.Core().V1().Services().Informer().AddEventHandler(eh)
 
+	informers := []cache.SharedIndexInformer{
+		coreInformers.Core().V1().Services().Informer(),
+		contourInformers.Contour().V1beta1().IngressRoutes().Informer(),
+		contourInformers.Contour().V1beta1().TLSCertificateDelegations().Informer(),
+		contourInformers.Projectcontour().V1().HTTPProxies().Informer(),
+		contourInformers.Projectcontour().V1().TLSCertificateDelegations().Informer(),
+	}
+
 	// After K8s 1.13 the API server will automatically translate extensions/v1beta1.Ingress objects
 	// to networking/v1beta1.Ingress objects so we should only listen for one type or the other.
 	// The default behavior is to listen for networking/v1beta1.Ingress objects and let the API server
 	// transparently upgrade the extensions version for us.
 	if ctx.UseExtensionsV1beta1Ingress {
-		coreInformers.Extensions().V1beta1().Ingresses().Informer().AddEventHandler(eh)
+		informers = append(informers, coreInformers.Extensions().V1beta1().Ingresses().Informer())
 	} else {
-		coreInformers.Networking().V1beta1().Ingresses().Informer().AddEventHandler(eh)
+		informers = append(informers, coreInformers.Networking().V1beta1().Ingresses().Informer())
 	}
-	contourInformers.Contour().V1beta1().IngressRoutes().Informer().AddEventHandler(eh)
-	contourInformers.Contour().V1beta1().TLSCertificateDelegations().Informer().AddEventHandler(eh)
-	contourInformers.Projectcontour().V1().HTTPProxies().Informer().AddEventHandler(eh)
-	contourInformers.Projectcontour().V1().TLSCertificateDelegations().Informer().AddEventHandler(eh)
 
 	// Add informers for each root-ingressroute namespaces
 	for _, inf := range namespacedInformers {
-		inf.Core().V1().Secrets().Informer().AddEventHandler(eh)
+		informers = append(informers, inf.Core().V1().Secrets().Informer())
 	}
 	// If root-ingressroutes are not defined, then add the informer for all namespaces
 	if len(namespacedInformers) == 0 {
-		coreInformers.Core().V1().Secrets().Informer().AddEventHandler(eh)
+		informers = append(informers, coreInformers.Core().V1().Secrets().Informer())
```

Comment:
Maybe a small helper that took an informer and a resource handler and returned the informer would help
```
informers = append(informers, registerEventHandler(inf.Core().V1().Secrets().Informer())
...
func registerEventHandler(inf <informer type>, eh cache.ResourceEventHandler) <informer type> {
      inf.AddEventHandler(eh)
      return inf
}
```
Then every time we add an informer to the slice, we can also register an event handler at the same time